output/sarif.go

```diff
@@ -0,0 +1,496 @@
+		rule.Properties["rule"] = ruleName
+	}
+	if query, ok := result.Metadata["query"].(string); ok {
+		rule.Properties["query"] = query
+	}
+	rule.Properties["namespace"] = namespace
+
+	// Add additional rule metadata if available
+	if desc, ok := result.Metadata["description"].(string); ok {
+		rule.FullDescription = &sarifMessage{Text: desc}
+	}
+	if url, ok := result.Metadata["url"].(string); ok {
+		rule.HelpURI = url
+	}
+	if help, ok := result.Metadata["help"].(string); ok {
+		rule.Help = &sarifMessage{Text: help}
+	}
+
+	// Add any remaining metadata to properties
+	for k, v := range result.Metadata {
+		switch k {
+		case "package", "rule", "description", "url", "help", "query":
+			// Skip already processed fields
+			continue
+		default:
+			rule.Properties[k] = v
+		}
+	}
+
+	return rule
+}
+
+// createLocation creates a new SARIF location from a file path
+func createLocation(filePath string) sarifLocation {
+	return sarifLocation{
+		PhysicalLocation: sarifPhysicalLocation{
+			ArtifactLocation: sarifArtifactLocation{
+				URI: filepath.ToSlash(filePath),
+			},
+		},
+	}
+}
+
+// createProperties creates a new properties map with namespace information
+func createProperties(metadata map[string]interface{}, namespace string) map[string]interface{} {
+	properties := make(map[string]interface{})
+	for k, v := range metadata {
+		switch k {
+		case "package", "rule", "description", "url", "help":
+			// Skip rule-level metadata
+			continue
+		case "query", "traces", "outputs":
+			// Include query-related information
+			properties[k] = v
+		default:
+			properties[k] = v
+		}
+	}
+
+	// Always include namespace
+	properties["namespace"] = namespace
+
+	return properties
+}
+
+// processResults processes a slice of results and adds them to the SARIF run
+func processResults(run *sarifRun, results []Result, rType resultType, fileName, namespace string, ruleMap map[string]bool) error {
+	for _, result := range results {
+		// Create or get rule
+		rule := createRule(result, rType, namespace)
+		if !ruleMap[rule.ID] {
+			run.Tool.Driver.Rules = append(run.Tool.Driver.Rules, rule)
+			ruleMap[rule.ID] = true
+		}
+
+		// Find rule index
+		ruleIndex := -1
+		for i, r := range run.Tool.Driver.Rules {
+			if r.ID == rule.ID {
+				ruleIndex = i
+				break
+			}
+		}
+
+		if ruleIndex == -1 {
+			return fmt.Errorf("rule %s not found in rules array after being added", rule.ID)
+		}
+
+		// Create result
+		run.Results = append(run.Results, sarifResult{
+			RuleID:     rule.ID,
+			RuleIndex:  ruleIndex,
+			Kind:       rType.kindStr,
+			Level:      rType.level,
+			Message:    sarifMessage{Text: result.Message},
+			Locations:  []sarifLocation{createLocation(fileName)},
+			Properties: createProperties(result.Metadata, namespace),
+		})
+	}
+	return nil
+}
+
+// createSuccessResult creates a success result for a given file and namespace
+func createSuccessResult(run *sarifRun, fileName, namespace string, ruleMap map[string]bool) error {
```

Comment:
Please update this to have an explicit return rather than having an input parameter also be used as an output.